snafu/benchmarks/nighthawk/nighthawk.py

```diff
@@ -0,0 +1,272 @@
+            "--duration", 
+            dest="duration", 
+            env_var="DURATION", 
+            help="The number of seconds that the test should run", 
+            default=60, 
+            type=int),
+        ConfigArgument(
+            "--concurrency", 
+            dest="concurrency", 
+            env_var="CONCURRENCY", 
+            help="The number of concurrent event loops that should be used. Specify 'auto' to "\
+                "let Nighthawk leverage all vCPUs that have affinity to the Nighthawk process"\
+                ". Note that increasing this results in an effective load multiplier combined"\
+                " with the configured --rps and --connections values", 
+            default=1, 
+            type=int),
+    )
+
+    def setup(self) -> bool:
+        """Parse config and check for validations."""
+        self.config.parse_args()
+        self.logger.debug(f"Got config: {vars(self.config)}")
+
+        if not getattr(self.config, "user", False) or not getattr(self.config, "uuid", False):
+            self.logger.critical("Missing required metadata. Need both user and uuid to continue")
+            return False
+
+        return True
+
+    def _parse_stdout(self) -> NighthawkStat:
+        """
+        Return parsed stdout of Nighthawk sample.
+
+        Returns
+        -------
+        NighthawkStat
+        """
+
+        data = json.load(open("nighthawk.json"))
+        # populating latency in milliseconds and throughput as queries per second.
+        latency_percentiles = {}
+        duration_histogram = data['DurationHistogram']
+        for each_percentile in duration_histogram['Percentiles']:
+            percentile = str(each_percentile['Percentile'])
+            if percentile not in latency_percentiles.keys():
+                latency_percentiles[percentile] = 0
+            latency_percentiles[percentile] += each_percentile['Value'] * 1000
+
+        return NighthawkStat(
+            requested_qps=data['RequestedQPS'],
+            actual_qps=data["ActualQPS"], 
+            throughput=data["ActualQPS"],
+            status_codes=data["RetCodes"],
+            p50_latency=latency_percentiles.get("50", None),
+            p75_latency=latency_percentiles.get("75", None),
+            p80_latency=latency_percentiles.get("80", None),
+            p90_latency=latency_percentiles.get("90", None),
+            p95_latency=latency_percentiles.get("95", None),
+            p99_latency=latency_percentiles.get("99", None),
+            p99_9_latency=latency_percentiles.get("99.9", None),
+            avg_latency=duration_histogram['Avg'] * 1000,
+            timestamp=data['StartTime'],
+            bytes_in=float(data['BytesReceived']),
+            bytes_out=float(data['BytesSent'])
+        )
+
+    def _run_nighthawk(self):
+        """
+        Method to execute nighthawk command.
+        """
+
+        cmd = (
+            "nighthawk_client --concurrency {0} --duration {1} --connections {2} "
```

Comment:
Do we have data on using the `simple-warmup` function?

Reply:
Yes, the excel sheet has the data for three iterations. The results are consistent for each run.